### PR TITLE
pkp/plagiarism#52 Plugin update with new iThenticate API integration

### PR DESCRIPTION
**Pull request by touhidurabir**
for pkp/plagiarism#52

This PR make the plugin compatible with `OJS/OMP/OPS` `stable-3.4.0-6` .

### CODE REVIEW

File: PlagiarismSubmissionSubmitListener.php [hunk 1]
```diff
@@ -3,8 +3,8 @@
 /**
  * @file plugins/generic/plagiarism/PlagiarismSubmissionSubmitListener.php
```
Comment (asmecher) [inline PlagiarismSubmissionSubmitListener.php:4]: Looking at the `diff`, this file appears to have DOS line endings; please convert over to UNIX-style.

File: TestIthenticate.php [hunk 2]
```diff
@@ -1,49 +1,473 @@
...
+        ?string $eulaVersion = null
+    )
+    {
+        error_log(
+            sprintf(
+                "Constructing iThenticate with API URL : {$apiUrl} \n
+                API Key : {$apiKey} \n
+                Integration Name : {$integrationName} \n
+                Integration Version : {$integrationVersion} \n 
+                EUlA Version : {$eulaVersion}"
```
Comment (asmecher) [inline TestIthenticate.php:133]: Typo in case: `EUlA` should be `EULA`

File: controllers/PlagiarismArticleGalleyGridHandler.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	public function initialize($request, $args = null)
+	{
+		parent::initialize($request, $args);
+		
+		$this->addColumn(new SimilarityActionGridColumn(static::$_plugin));
+	}
+}
+
+if (!PKP_STRICT_MODE) {
+    class_alias('\APP\plugins\generic\plagiarism\controllers\PlagiarismArticleGalleyGridHandler', '\PlagiarismArticleGalleyGridHandler');
```
Comment (asmecher) [inline controllers/PlagiarismArticleGalleyGridHandler.php:59]: I don't think this `class_alias` should be necessary.

File: controllers/PlagiarismComponentHandler.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	{
+		$this->addPolicy(new UserRequiredPolicy($request));
+		$this->addPolicy(new UserRolesRequiredPolicy($request));
+		
+		return parent::authorize($request, $args, $roleAssignments);
+	}
+}
+
+if (!PKP_STRICT_MODE) {
+    class_alias('\APP\plugins\generic\plagiarism\controllers\PlagiarismComponentHandler', '\PlagiarismComponentHandler');
```
Comment (asmecher) [inline controllers/PlagiarismComponentHandler.php:60]: I don't think this should be necessary.

File: controllers/PlagiarismIthenticateActionHandler.php [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+			}
+		}
+
+		return DAO::getDataChangedEvent($submissionFile->getId());
+	}
+
+}
+
+if (!PKP_STRICT_MODE) {
+    class_alias('\APP\plugins\generic\plagiarism\controllers\PlagiarismIthenticateActionHandler', '\PlagiarismIthenticateActionHandler');
```
Comment (asmecher) [inline controllers/PlagiarismIthenticateActionHandler.php:468]: I don't think this should be necessary.

File: controllers/PlagiarismWebhookHandler.php [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+	{
+		return Capsule::table(Repo::submissionFile()->getCollector()->dao->settingsTable)
+			->where('setting_name', 'ithenticateId')
+			->where('setting_value', $id)
+			->first();
+	}
+}
+
+if (!PKP_STRICT_MODE) {
+    class_alias('\APP\plugins\generic\plagiarism\controllers\PlagiarismWebhookHandler', '\PlagiarismWebhookHandler');
```
Comment (asmecher) [inline controllers/PlagiarismWebhookHandler.php:242]: I don't think this should be necessary.

File: grids/RearrangeColumnsFeature.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+			+ [SimilarityActionGridColumn::SIMILARITY_ACTION_GRID_COLUMN_ID => $plagiarismColumn]
+			+ $this->gridHandler->_columns;
+
+		// set the first file name column as first column
+		reset($this->gridHandler->_columns)->addFlag('firstColumn', true);
+	}
+}
+
+if (!PKP_STRICT_MODE) {
+    class_alias('\APP\plugins\generic\plagiarism\grids\RearrangeColumnsFeature', '\RearrangeColumnsFeature');
```
Comment (asmecher) [inline grids/RearrangeColumnsFeature.php:64]: I don't think this should be necessary.

File: grids/SimilarityActionGridColumn.php [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+			return WORKFLOW_STAGE_ID_PRODUCTION;
+		}
+
+		return $request->getUserVar('stageId');
+	}
+
+}
+
+if (!PKP_STRICT_MODE) {
+    class_alias('\APP\plugins\generic\plagiarism\grids\SimilarityActionGridColumn', '\SimilarityActionGridColumn');
```
Comment (asmecher) [inline grids/SimilarityActionGridColumn.php:369]: I don't think this should be necessary.

File: templates/similarityScore.tpl [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        target="_blank"
+        href="{$viewerUrl}"
+        title="{translate key="plugins.generic.plagiarism.similarity.action.launch.viewer.title"}"
+    >
+        <img 
+            src="{$logoUrl}"
+            alt="{translate key="plugins.generic.plagiarism.similarity.match.title"}"
+        />
+    </a>
+    <span>{$score}%</span>
```
Comment (asmecher) [inline templates/similarityScore.tpl:22]: Should probably have an `|escape` modifier, for safety's sake.

File: tools/registerWebhooks.php [hunk 10]
```diff
@@ -0,0 +1,134 @@
+<?php
```
Comment (asmecher) [inline tools/registerWebhooks.php:1]: This file uses tab-based indents; we should be using spaces.

**Comment by asmecher** [comment]
Thanks, @touhidurabir, I looked over a diff between the 3.3.x and proposed 3.4.x implementations and it looks clean. I didn't test it, though -- let's walk through that when you're ready.

File: controllers/PlagiarismIthenticateActionHandler.php (hunk 5 above)
Comment (touhidurabir) [reply controllers/PlagiarismIthenticateActionHandler.php:474]: @asmecher Thats how it should be but when calling these controllers as part of hook `LoadComponentHandler`, in the `PlagiarismPlugin::handleRouteComponent`, I am getting the following exception without the class alias. any idea what I am missing ? 

```
[11-Sep-2024 16:01:03 Asia/Dhaka] PHP Fatal error:  Uncaught Exception: Cannot instantiate class. Class "PlagiarismIthenticateActionHandler" is not declared in "plugins.generic.plagiarism.controllers.PlagiarismIthenticateActionHandler". in /Users/abir/Sites/code/ojs-stable-3_4_0/lib/pkp/includes/functions.php:236
Stack trace:
#0 /Users/abir/Sites/code/ojs-stable-3_4_0/lib/pkp/classes/core/PKPComponentRouter.php(252): instantiate('plugins.generic...', 'PKPHandler', NULL, Array)
#1 /Users/abir/Sites/code/ojs-stable-3_4_0/lib/pkp/classes/core/PKPComponentRouter.php(279): PKP\core\PKPComponentRouter->getRpcServiceEndpoint(Object(APP\core\Request))
#2 /Users/abir/Sites/code/ojs-stable-3_4_0/lib/pkp/classes/core/Dispatcher.php(165): PKP\core\PKPComponentRouter->route(Object(APP\core\Request))
#3 /Users/abir/Sites/code/ojs-stable-3_4_0/lib/pkp/classes/core/PKPApplication.php(395): PKP\core\Dispatcher->dispatch(Object(APP\core\Request))
#4 /Users/abir/Sites/code/ojs-stable-3_4_0/index.php(21): PKP\core\PKPApplication->execute()
#5 /Users/abir/.composer/vendor/laravel/valet/server.php(110): require('/Users/abir/Sit...')
#6 {main}
  thrown in /Users/abir/Sites/code/ojs-stable-3_4_0/lib/pkp/includes/functions.php on line 236
```

Comment (asmecher) [reply controllers/PlagiarismIthenticateActionHandler.php:473]: @touhidurabir Have a look at point 6 under "Additional details on plugins" in https://github.com/pkp/pkp-lib/issues/6091#issuecomment-796314189 -- you can look at the [static pages plugin](https://github.com/pkp/staticPages) for an example. Basically starting with 3.4.0 the hook callback function can now provide an instance of a handler class directly.

**Comment by touhidurabir** [comment]
@asmecher all updated and tested for OMP and OPS, seems working fine . should ready for a final look or better, if we can have a test instance for 3.4.0 to test it .

**Pull request merged**